README.md
# Turkpin API Package

[![Latest Version on Packagist](https://img.shields.io/packagist/v/anilken/turkpin-api.svg?style=flat-square)](https://packagist.org/packages/anilken/turkpin-api)
[![Total Downloads](https://img.shields.io/packagist/dt/anilken/turkpin-api.svg?style=flat-square)](https://packagist.org/packages/anilken/turkpin-api)

**Turkpin API** With this package, you can create an epin order and deliver it to your customers. **[Official Documentation](https://dev.turkpin.com/view/3051588/TVza8YYA#038621b3-851b-4ac9-82ef-a2ad22f7cd20)**

## Installation

You can install the package via composer:

```bash
composer require anilken/turkpin-api
```

## Usage

```php
use Anilken\Turkpin\Turkpin;

$api = new Turkpin('username', 'password');

$api->epinList();

// Example Output

[
  {
    "id": "5",
    "name": "Echo of Soul (EOS)"
  },
  {
    "id": "633",
    "name": "Google Play kodu"
  }
  ...
]


$api->epinProducts($epin_id);

// Example Output

[
  {
    "name": "Google Play kodu 25 TL",
    "id": "8209",
    "stock": "377",
    "min_order": "1",
    "max_order": "0",
    "price": "24.5000",
    "tax_type": []
  },
  ...
]

$api->epinOrder($epin_id, $product_id, $qty = 1, $character = '');

// Example Output

{
  "status": "Success",
  "order_no": "21061113513501",
  "total_amount": "1",
  "list": [
    {
      "code": "8PFG-XXXX-FL3C-XXXX",
      "desc": "Test Ürün"
    }
  ]
}

$api->checkStatus($order_no);

// Example Output

{
  "status_code": "000",
  "order_no": "21061113513501",
  "order_code": "2",
  "order_status_description": "Siparişiniz Tamamlandı",
  "check_date": "11-06-2021 13:51:53",
  "extra": "İPTAL SEBEBI",
}

$api->checkBalance();

// Example Output

{
  "balance": "996501.2300",
  "credit": "0.0000",
  "bonus": "0.000",
  "spending": "3498.7700"
}

```

## Exception Codes

| Code | Message                                                                                             |
|------|-----------------------------------------------------------------------------------------------------|
| 0	   | Empty api result try again                                                                          |
| 1	   | User not found, check your username and password.                                                   |
 | 2	   | There is no access from your IP address. Please check your IP Address.                              |
 | 3	   | Invalid command, check cmd parameter.                                                               |
 | 4	   | Message field is a mandatory field.                                                                 |
 | 5	   | An error occurred while processing.                                                                 |
 | 6	   | This Account is not authorized to use Turkpin API                                                   |
 | 7	   | This Account is not available for Turkpin API.                                                      |
 | 8	   | Game ID is missing                                                                                  |
 | 9	   | The server list for the game could not be found                                                     |
 | 10	  | Missing or incorrect order format.                                                                  |
 | 11	  | The product was not found.                                                                          |
 | 12	  | The product is out of stock.                                                                        |
 | 13	  | The number of orders must be min% s.                                                                |
 | 14	  | Dealer Balance Insufficient.                                                                        |
 | 15	  | Payment type is invalid                                                                             |
 | 16	  | Missing or incorrect notification format. Please check fields                                       |
 | 17	  | You have entered a missing or incorrect notification number. Please check the notification_id value |
 | 18	  | Number of orders must be max% s.                                                                    |
 | 19	  | System error, error code:% s                                                                        |
 | 20	  | Please enter correct operator and phone number for GSM TL top up                                    |
 | 21	  | The amount to be reported for payment must be greater than 0                                        |
 | 22	  | This product is not purchased by us.                                                                |
 | 23	  | The site is currently closed for an update or maintenance.                                          | 
 | 99   | Order Not Found. Check Order Number                                                                 |
 | 111  | XML Format Error. Please check the XML format you sent.                                             |
| ---  | Custom error codes added to this package                                                            |
| 101  | Failed to get game list                                                                             |
| 102  | Failed to get products list                                                                         |
| 103  | Failed to get balance                                                                               |

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](https://github.com/spatie/.github/blob/main/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [ANILKEN](https://github.com/anilken)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `turkpin-api` will be documented in this file.

## v1.0.6 - 2022-09-01

**Full Changelog**: https://github.com/anilken/turkpin-api/compare/v1.0.5...v1.0.6

## v1.0.5 - 2022-09-01

### What's Changed

- Bump dependabot/fetch-metadata from 1.3.1 to 1.3.3 by @dependabot in https://github.com/anilken/turkpin-api/pull/4

### New Contributors

- @dependabot made their first contribution in https://github.com/anilken/turkpin-api/pull/4

**Full Changelog**: https://github.com/anilken/turkpin-api/compare/v1.0.4...v1.0.5

## v1.0.4 - 2022-05-24

## What's Changed

- Api www Required by @anilken in https://github.com/anilken/turkpin-api/pull/3

**Full Changelog**: https://github.com/anilken/turkpin-api/compare/v1.0.3...v1.0.4

## v1.0.3 - 2022-05-20

## What's Changed

- Add check epin stock by @anilken in https://github.com/anilken/turkpin-api/pull/2

**Full Changelog**: https://github.com/anilken/turkpin-api/compare/v1.0.2...v1.0.3

## v1.0.2 - 2022-05-19

## What's Changed

- Live api fix by @anilken in https://github.com/anilken/turkpin-api/pull/1

**Full Changelog**: https://github.com/anilken/turkpin-api/compare/v1.0.1...v1.0.2

## v1.0.1 - 2022-05-19

**Full Changelog**: https://github.com/anilken/turkpin-api/compare/v1.0.0...v1.0.1

## v1.0.0 - 2022-05-11

Initial commit

**Full Changelog**: https://github.com/anilken/turkpin-api/commits/v1.0.0


src/Base.php
<?php

namespace Anilken\Turkpin;

use Anilken\Turkpin\Exceptions\Exception;
use SimpleXMLElement;

class Base
{
    protected $api_url = "https://www.turkpin.com/api.php";

    /**
     * @var string Turkpin Api Username
     */
    protected $username;

    /**
     * @var string Turkpin Api Password
     */
    protected $password;

    /**
     * @param $username
     * @param $password
     * @param bool $is_test
     */
    public function __construct($username, $password, $is_test = false)
    {
        $this->username = $username;
        $this->password = $password;

        if ($is_test === true) {
            $this->api_url = "https://www.turkpin.net/api.php";
        }
    }

    /**
     * @param $cmd
     * @param array|null $extra
     * @return mixed
     * @throws Exception
     */
    public function make($cmd, ?array $extra = [])
    {
        $api_data = [
            'params' => [
                'username' => $this->username,
                'password' => $this->password,
                'cmd' => $cmd,
            ],
        ];

        foreach ($extra as $k => $v) {
            $api_data['params'][$k] = $v;
        }


        $xml = new SimpleXMLElement('<APIRequest/>');

        $post = $this->array2xml($api_data, $xml)->asXML();

        return $this->request($post);
    }

    /**
     * @param $xml
     * @return mixed
     * @throws Exception
     */
    protected function request($xml)
    {
        $curl = curl_init();

        curl_setopt_array($curl, [
            CURLOPT_URL => $this->api_url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_SSL_VERIFYPEER => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => ['DATA' => $xml],
        ]);

        $response = curl_exec($curl);

        curl_close($curl);

        if (! $response || strlen(trim($response)) == 0) {
            throw new Exception('Api empty result', 0);
        }

        $result = $this->xml2array($response)['params'];

        if (isset($result['HATA_NO']) && $result['HATA_NO'] !== '000') {
            throw new Exception($result['HATA_ACIKLAMA'], $result['HATA_NO']);
        }

        if (isset($result['error']) && $result['error'] !== '000') {
            throw new Exception($result['error_desc'], $result['error']);
        }

        return $result;
    }

    /**
     * @param array $data
     * @param SimpleXMLElement $xml
     * @return SimpleXMLElement
     */
    protected function array2xml(array $data, SimpleXMLElement $xml): SimpleXMLElement
    {
        foreach ($data as $k => $v) {
            is_array($v) ? $this->array2xml($v, $xml->addChild($k)) : $xml->addChild($k, $v);
        }

        return $xml;
    }

    /**
     * @param $xml_data
     * @return mixed
     */
    protected function xml2array($xml_data)
    {
        $xml = simplexml_load_string($xml_data, "SimpleXMLElement", LIBXML_NOCDATA);

        return json_decode(json_encode($xml), true);
    }
}


src/Turkpin.php
<?php

namespace Anilken\Turkpin;

use Anilken\Turkpin\Exceptions\Exception;
use Anilken\Turkpin\Exceptions\InvalidArgumentException;

class Turkpin
{
    protected $api;

    /**
     * Turkpin constructor
     * Set username password for api
     *
     * @param string $username Turkpin Api Username
     * @param string $password Turkpin Api Password
     * @throws Exception
     * @throws InvalidArgumentException
     */
    public function __construct($username, $password, $is_test = false)
    {
        if (! function_exists('curl_version')) {
            throw new Exception('CURL not installed');
        }

        $this->api = new Base($username, $password, $is_test);
    }

    /**
     * Get List of games
     *
     * @return array
     * @throws Exception
     */
    public function epinList(): array
    {
        $result = $this->api->make('epinOyunListesi');

        if (! isset($result['oyunListesi']['oyun'])) {
            throw new Exception('Failed to get game list', 101);
        }

        return $result['oyunListesi']['oyun'];
    }

    /**
     * Get List of products
     *
     * @param $epin_id
     * @return array
     * @throws Exception
     */
    public function epinProducts($epin_id): array
    {
        $result = $this->api->make('epinUrunleri', ['oyunKodu' => $epin_id]);

        if (! isset($result['epinUrunListesi']['urun'])) {
            throw new Exception('Failed to get products list', 102);
        }

        return $result['epinUrunListesi']['urun'];
    }

    /**
     * Create new order
     *
     * @param int $epin_id epinList() id
     * @param int $product_id epinProducts() id
     * @param int $qty order qty
     * @param string $character optional character to be delivered
     * @return array
     * @throws Exception
     */
    public function epinOrder($epin_id, $product_id, $qty = 1, $character = ''): array
    {
        $result = $this->api->make('epinSiparisYarat', ['oyunKodu' => $epin_id, 'urunKodu' => $product_id, 'adet' => $qty, 'character' => $character]);

        $r = [
            'status' => $result['siparisSonuc'],
            'order_no' => $result['siparisNo'],
            'total_amount' => $result['siparisTutari'],
        ];


        if (isset($result['epin_list']['epin'])) {
            $r['list'] = $result['epin_list']['epin'];
        }

        //fix array
        if (isset($result['epin_list']['epin']) && isset($result['epin_list']['epin']['code']) == 1) {
            $r['list'] = [$result['epin_list']['epin']];
        }

        return $r;
    }

    /**
     * Check in stock product
     *
     * @param int $epin_id epinList() id
     * @param int $product_id epinProducts() id
     * @return int
     * @throws Exception
     */
    public function epinCheckStock($epin_id, $product_id): int
    {
        $products = $this->epinProducts($epin_id);

        $stock = 0;
        foreach ($products as $product) {
            if ($product['id'] === $product_id) {
                $stock = $product['stock'];

                break;
            }
        }

        return $stock;
    }

    /**
     * Check order Status
     *
     * @param int|string $order_no epinOrder() order_no
     * @return array
     * @throws Exception
     */
    public function checkStatus($order_no): array
    {
        $result = $this->api->make('siparisDurumu', ['siparisNo' => $order_no]);

        $r = [
            'status_code' => $result['DURUM_KODU'],
            'order_no' => $result['SIPARIS_NO'],
            'order_code' => $result['SIPARIS_DURUMU'],
            'order_status_description' => $result['SIPARIS_DURUMU_ACIKLAMA'],
            'check_date' => $result['KONTROL_TARIHI'],
        ];

        if (isset($result['EKSTRA'])) {
            $r['extra'] = $result['EKSTRA'];
        }

        return $r;
    }

    /**
     * Check account balance
     *
     * @return mixed
     * @throws Exception
     */
    public function checkBalance()
    {
        $result = $this->api->make('balance');

        if (! isset($result['balanceInformation'])) {
            throw new Exception('Failed to get balance', 103);
        }

        return $result['balanceInformation'];
    }
}
